### PR TITLE
DPO example with gemma3

### PR DESCRIPTION
**Pull request by immuntasir**
This pull request introduces a DPO (Direct Preference Optimization) example using the Gemma3 model. Key changes include:

**DPO Example Notebook**: Added a comprehensive Colab Notebook (dpo_demo_gemma3.ipynb) demonstrating the DPO training process with the Gemma3 model. The notebook includes steps for loading the model, preparing datasets, training with DPO, and evaluating the model.

Gemma3Tokenizer: Added a new Gemma3Tokenizer class in translation_dataset.py.

### CODE REVIEW

File: examples/dpo_demo_gemma3.ipynb [hunk 1]
```diff
@@ -0,0 +1,849 @@
...
+    "from tunix.examples.data import translation_dataset as data_lib\n",
+    "from tunix.generate import sampler as sampler_lib\n",
+    "from tunix.models.gemma3 import params_safetensors as params_safetensors_lib\n",
+    "\n",
+    "from datasets import load_dataset\n",
+    "from tunix.sft.dpo.dpo_trainer import DpoTrainingConfig\n",
+    "from tunix.sft.dpo.dpo_trainer import DpoTrainer\n",
+    "from tunix.sft.dpo.dpo_trainer import TrainingInput\n",
+    "from huggingface_hub import snapshot_download\n",
+    "from tunix.sft.dpo.dpo_trainer import _generate_ids_and_masks\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Do we need this? It's a private function. Can we use `process_dpo_record` directly?

File: examples/dpo_demo_gemma3.ipynb [hunk 2]
```diff
@@ -0,0 +1,849 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!pip install -q kagglehub\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: We install `kagglehub` here, but use `huggingface_hub` below. Is it possible to use `kagglehub` throughout? Or are there any issues?

File: examples/dpo_demo_gemma3.ipynb [hunk 3]
```diff
@@ -0,0 +1,849 @@
+{
+ "cells": [
+  {
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Do you want to give a brief description at the top, as to what the purpose of the example is, and what hardware you are running it on? Might be useful for the user! :)

File: examples/dpo_demo_gemma3.ipynb [hunk 4]
```diff
@@ -0,0 +1,849 @@
...
+    "!huggingface-cli login"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "model_id = \"google/gemma-3-1b-it\"\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Maybe set this at the top, i.e., the place where we've defined all hyperparams?

File: examples/dpo_demo_gemma3.ipynb [hunk 5]
```diff
@@ -0,0 +1,849 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def show_hbm_usage():\n",
+    "  \"\"\"Displays memory usage per device.\"\"\"\n",
+    "  fmt_size = functools.partial(humanize.naturalsize, binary=True)\n",
+    "\n",
+    "  print(\"\\n--- TPU HBM Usage ---\")\n",
+    "  for i, d in enumerate(jax.local_devices()):\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Maybe just import this from here? 

https://github.com/google/tunix/blob/e03f154edd2abfddd6b9babb72a3d0d3c4d81bb2/tunix/rl/utils.py#L149

File: examples/dpo_demo_gemma3.ipynb [hunk 6]
```diff
@@ -0,0 +1,849 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "MODEL_CP_PATH = local_model_path\n",
+    "\n",
+    "model_config = (\n",
+    "    gemma3_model_lib.Gemma3Config.gemma3_1b()\n",
+    ")  # pick correponding config based on model version\n",
+    "MESH = [(1, 1), (\"fsdp\", \"tp\")]\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: MESH Is defined above in the config cell. In general, let's try to use stuff from the config cell, and remove vars which are not being used

File: examples/dpo_demo_gemma3.ipynb [hunk 7]
```diff
@@ -0,0 +1,849 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
+    "\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Remove?

File: examples/dpo_demo_gemma3.ipynb [hunk 8]
```diff
@@ -0,0 +1,849 @@
...
+    "gemma_tokenizer = data_lib.Gemma3Tokenizer()\n",
+    "#from transformers import AutoTokenizer\n",
+    "\n",
+    "#gemma_tokenizer = AutoTokenizer.from_pretrained(\"google/gemma-3-1b-it\")\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Remove?

File: examples/dpo_demo_gemma3.ipynb [hunk 9]
```diff
@@ -0,0 +1,849 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
+    "\n",
+    "gemma_tokenizer = data_lib.Gemma3Tokenizer()\n",
+    "#from transformers import AutoTokenizer\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Remove?

File: examples/dpo_demo_gemma3.ipynb [hunk 10]
```diff
@@ -0,0 +1,849 @@
...
+    "Load evaluation data and evaluate the reference model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "eval_dataset = load_dataset(\"gsm8k\", \"main\", split=\"test\").select(range(NUM_TEST_BATCHES))"
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Was there an issue with using TFDS, like all other examples use? Maybe we should use that for uniformity

File: examples/dpo_demo_gemma3.ipynb [hunk 11]
```diff
@@ -0,0 +1,849 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "reasoning_start = \"<reasoning>\"\n",
+    "reasoning_end = \"</reasoning>\"\n",
+    "solution_start = \"<answer>\"\n",
+    "solution_end = \"</answer>\"\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Confused why it's not showing up in the notebook?

image

File: examples/dpo_demo_gemma3.ipynb [hunk 12]
```diff
@@ -0,0 +1,849 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Let's add a heading here:  "# Install necessary libraries"

File: examples/dpo_demo_gemma3.ipynb [hunk 13]
```diff
@@ -0,0 +1,849 @@
...
+    "!pip install -q huggingface_hub\n",
+    "!pip install -q datasets"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Markdown heading here: "# Imports"

File: examples/dpo_demo_gemma3.ipynb [hunk 14]
```diff
@@ -0,0 +1,849 @@
...
+    "import numpy as np"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# ====== Data ======\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Heading here: "# Hyperparamters/Config"

File: examples/dpo_demo_gemma3.ipynb [hunk 15]
```diff
@@ -0,0 +1,849 @@
...
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "hnxxLtLAv66G"
+   },
+   "source": [
+    "# Load reference model and LoRA model"
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: An explanation here would be nice...can talk about the reference model staying fixed during training, and the LoRA model is the policy model, which is trained, reference model being a Gemma 3 1B model, etc.

File: examples/dpo_demo_gemma3.ipynb [hunk 16]
```diff
@@ -0,0 +1,849 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!huggingface-cli login"
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: @tianshub - do we want to stick with Kaggle throughout, or are we okay with using HF?

File: examples/dpo_demo_gemma3.ipynb [hunk 17]
```diff
@@ -0,0 +1,849 @@
...
+    "match_numbers.findall(f\"{solution_start}  0.34  {solution_end}\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def evaluate(\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Do you want to talk about what accuracy we get here? Roughly. We can mention it is around xy.z% or something

File: examples/dpo_demo_gemma3.ipynb [hunk 18]
```diff
@@ -0,0 +1,849 @@
...
+    "dpo_dataset = load_dataset(\"argilla/distilabel-intel-orca-dpo-pairs\", split=\"train\")\n",
+    "gsm8k_train_dpo_dataset = dpo_dataset.filter(lambda x: x['in_gsm8k_train'])\n",
+    "\n",
+    "# Get the number of samples in the filtered dataset\n",
+    "num_gsm8k_train_samples = len(gsm8k_train_dpo_dataset)\n",
+    "print(f\"Number of samples with in_gsm8k_train=True: {num_gsm8k_train_samples}\")\n",
+    "\n",
+    "# Calculate how many more samples are needed\n",
+    "total_samples_needed = NUM_BATCHES\n",
+    "samples_to_add = total_samples_needed - num_gsm8k_train_samples\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Confused about this...this will work for batch size 1, but won't work for batch size > 1, right?

`gsm8k_train_dpo_dataset` is not batched above, maybe we should batch it up first?

File: examples/dpo_demo_gemma3.ipynb [hunk 19]
```diff
@@ -0,0 +1,849 @@
...
+    "    # Ensure we don't sample more than the total available in the original dataset\n",
+    "    random_samples = dpo_dataset.shuffle(seed=42).select(range(min(samples_to_add, len(dpo_dataset))))\n",
+    "    print(f\"Number of random samples selected: {len(random_samples)}\")\n",
+    "\n",
+    "    # Combine the filtered dataset and the random samples\n",
+    "    combined_dpo_dataset = concatenate_datasets([gsm8k_train_dpo_dataset, random_samples])\n",
+    "else:\n",
+    "    combined_dpo_dataset = gsm8k_train_dpo_dataset\n",
+    "\n",
+    "print(f\"Total samples in the combined dataset: {len(combined_dpo_dataset)}\")\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Okay, so here, we keep all samples present in GSM8K and add a few non-GSM8K samples randomly, right?

Good to add that we are trying to do the above here, and also explain the purpose of it. We can add it to a markdown cell

File: examples/dpo_demo_gemma3.ipynb [hunk 20]
```diff
@@ -0,0 +1,849 @@
...
+    "print(combined_dpo_dataset)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def process_dpo_dataset(dataset, tokenizer, max_prompt_length, total_generation_steps, batch_size):\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Can we just import `process_dpo_dataset` present in `dpo_trainer.py` directly?

File: examples/dpo_demo_gemma3.ipynb [hunk 21]
```diff
@@ -0,0 +1,849 @@
...
+    "):\n",
+    "  return grain.MapDataset.source(source).map(\n",
+    "      lambda x: TrainingInput(\n",
+    "          prompt_ids=prompt_ids[x],\n",
+    "          prompt_mask=prompt_mask[x],\n",
+    "          chosen_ids=chosen_ids[x],\n",
+    "          chosen_mask=chosen_mask[x],\n",
+    "          rejected_ids=rejected_ids[x],\n",
+    "          rejected_mask=rejected_mask[x],\n",
+    "      )\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: I believe a lot of the preprocessing code can be simplified here. Refer to this: https://github.com/google/tunix/blob/main/examples/dpo_demo.ipynb

File: examples/dpo_demo_gemma3.ipynb [hunk 22]
```diff
@@ -0,0 +1,849 @@
...
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "dpo_config"
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Move to previous cell

File: examples/dpo_demo_gemma3.ipynb [hunk 23]
```diff
@@ -0,0 +1,849 @@
...
+    "  )"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Configure DPO Training (using previously defined config variables)\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Can remove "(using previously defined config variables)"

File: examples/dpo_demo_gemma3.ipynb [hunk 24]
```diff
@@ -0,0 +1,849 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Start training\n",
+    "print(\"Starting DPO training...\")\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Can remove

File: examples/dpo_demo_gemma3.ipynb [hunk 25]
```diff
@@ -0,0 +1,849 @@
...
+    "\n",
+    "dpo_trainer.train(train_ds=processed_dpo_dataset)\n",
+    "print(\"DPO training finished.\")"
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Can remove

File: examples/dpo_demo_gemma3.ipynb [hunk 26]
```diff
@@ -0,0 +1,849 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: Markdown cell with heading: "# Evaluate". Can mention rough numbers, to show what kind of improvement we got

File: examples/dpo_demo_gemma3.ipynb [hunk 27]
```diff
@@ -0,0 +1,849 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "(corr, total, accuracy, partial_accuracy, format_accuracy), responses = evaluate(\n",
+    "    test_dataset,\n",
+    "    sampler,\n",
+    "    **GENERATION_CONFIGS[\"standard\"],\n",
+    "    make_lst=True\n",
+    ")\n",
```
Comment (abheesht17) [inline examples/dpo_demo_gemma3.ipynb]: You'd want to use `num_passes = 5` here, since you are using the "standard" gen config (and not greedy).

Alternatively, you can just use greedy generation here with `num_passes = 1`.

File: tunix/examples/data/translation_dataset.py [hunk 28]
```diff
@@ -89,11 +89,50 @@ def tokenize(
     return np.array(int_list, dtype=np.int32)
 
 
+class Gemma3Tokenizer(spm.SentencePieceProcessor):
```
Comment (abheesht17) [inline tunix/examples/data/translation_dataset.py]: Do we need this? Can't we just use the GemmaTokenizer and pass the correct SPM file path?

**Review by abheesht17 — CHANGES_REQUESTED** [review]
Thanks, awesome work on this! Left some nits and comments

File: examples/dpo_demo_gemma3.ipynb (hunk 16 above)
Comment (abheesht17) [reply examples/dpo_demo_gemma3.ipynb]: Chatted with Tianshu, okay to stick with HuggingFace in this notebook

File: examples/dpo_demo_gemma3.ipynb (hunk 18 above)
Comment (immuntasir) [reply examples/dpo_demo_gemma3.ipynb]: Up until this point, the dpo_dataset is in the HuggingFace datasets format, which is equivalent to a list of samples.

I was getting memory errors in Colab, so I am using a batch size of 1. If we want to add batching, we can probably add it to the process_dpo_dataset function.

File: examples/dpo_demo_gemma3.ipynb (hunk 19 above)
Comment (immuntasir) [reply examples/dpo_demo_gemma3.ipynb]: Yes, that is correct. Will add in a comment

File: examples/dpo_demo_gemma3.ipynb (hunk 21 above)
Comment (immuntasir) [reply examples/dpo_demo_gemma3.ipynb]: Thanks for pointing to this. I updated the notebook.

**Comment by abheesht17** [comment]
@immuntasir, could you please format your code by running https://github.com/google/tunix/blob/main/CONTRIBUTING.md#formatting-and-linting?

**Review by lc5211 — COMMENTED** [review]
Seems the notebook is malformed and blocked from internal reviews.

**Comment by immuntasir** [comment]
Hi @abheesht17 @lc5211 , I have run the pre-commit and updated the PR. Thank you.

**Comment by abheesht17** [comment]
For some reason, internally, the notebook shows up as malformed. I'm not sure why.

```
missing required field id in {...}->"cells"->[...]->{...}
```

Are you using Colab for this? If not, can you copy code from your Jupyter notebook to Colab, save, download and upload that notebook? Might work, but not sure!

**Comment by immuntasir** [comment]
Can you please check again? I have uploaded and downloaded from Colab, run pre-commit, and then pushed.
